### PR TITLE
cpu: matmul: enable optimized 8-bit gemm for PPC64

### PR DESCRIPTION
**Pull request by austinpagan**
# Description

Please include a summary of the change. Please also include relevant motivation and context. See [contribution guidelines](https://github.com/oneapi-src/oneDNN/blob/master/CONTRIBUTING.md) for more details. If the change fixes an issue not documented in the project's Github issue tracker, please document all steps necessary to reproduce it.

Greetings!  This change allows users running on the Power10 processor, with its "MMA engine," to access those instructions when performing eight bit matrix multiplications.  [we had previously enabled such functionality for bfloat16 as well as float32 and float64 in the OpenBLAS repository, but OpenBLAS does not support integer operations, so we are inserting that code here.
We noted that you wanted a short 50-character commit line, and we've done that.
We noted that you wanted the source code run through a "clang formatter" and we have also done that.
We noted that you want a pre-process if our new code changes the API or adds any new functions, but our code does neither, so we're hoping that proceeding with the pull request directly is ok...

Fixes # (github issue)

# Checklist

## General

- [ YES] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [ YES] Have you formatted the code using clang-format?

## Performance improvements

- [ N/A] Have you submitted performance data that demonstrates performance improvements?

### New features

- [ N/A] Have you published an RFC for the new feature?
- [ N/A] Was the RFC approved?
- [ N/A] Have you added relevant tests?

### Bug fixes

- [ N/A] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?
- [ N/A] Have you added relevant regression tests?

## [RFC](https://github.com/oneapi-src/oneDNN/tree/rfcs) PR

- [ N/A] Does RFC document follow the [template](https://github.com/oneapi-src/oneDNN/blob/rfcs/rfcs/template.md#onednn-design-document-rfc)?
- [ N/A] Have you added a link to the rendered document?

### CODE REVIEW

File: src/cpu/CMakeLists.txt [hunk 1]
```diff
@@ -38,6 +38,17 @@ if((DNNL_TARGET_ARCH STREQUAL "X64") OR (DNNL_TARGET_ARCH STREQUAL "AARCH64"))
     endforeach()
 endif()
 
+if(DNNL_TARGET_ARCH STREQUAL "PPC64")
+    file(GLOB FILES_REQUIRED_OPT
+        ${CMAKE_CURRENT_SOURCE_DIR}/gemm/*.[ch]pp
+        )
+    foreach(SOURCE_FILE ${FILES_REQUIRED_OPT})
+        list(APPEND SOURCES "${SOURCE_FILE}")
+    endforeach()
```
Comment (dzarukin) [inline src/cpu/CMakeLists.txt:47]: I feel this is a duplicate of line 21.

File: src/cpu/gemm/gemm.cpp [hunk 2]
```diff
@@ -30,6 +30,10 @@
 #include "cpu/gemm/f32/ref_gemm_f32.hpp"
 #include "cpu/gemm/s8x8s32/ref_gemm_s8x8s32.hpp"
 #include "cpu/gemm/s8x8s32/simple_gemm_s8s8s32.hpp"
+#ifdef DNNL_PPC64
+#include "cpu/ppc64/PPC64_gemm_driver.hpp"
+using namespace dnnl::impl::cpu;
+#endif
```
Comment (dzarukin) [inline src/cpu/gemm/gemm.cpp:36]: Please move this under:
```
#if DNNL_X64
...
#elif DNNL_PPC64
...
#endif
```

Please also do not use CamelCase for file names.

File: src/cpu/gemm/gemm.cpp [hunk 3]
```diff
@@ -192,6 +196,25 @@ dnnl_status_t gemm_s8x8s32(const char *transa, const char *transb,
                 B, LDB, bo, beta, C, LDC, co, false);
 #endif
 
+#ifdef DNNL_PPC64
+#ifdef __MMA__
+    if (!(utils::one_of(*transa, 'n', 'N', 't', 'T')
+                && utils::one_of(*transb, 'n', 'N', 't', 'T')))
+        return dnnl_unimplemented;
```
Comment (dzarukin) [inline src/cpu/gemm/gemm.cpp:203]: This part is likely belong to `check_gemm_x8x8x32_input(...)` function call. Could you move it there, please?

File: src/cpu/gemm/gemm.cpp [hunk 4]
```diff
@@ -192,6 +196,25 @@ dnnl_status_t gemm_s8x8s32(const char *transa, const char *transb,
                 B, LDB, bo, beta, C, LDC, co, false);
 #endif
 
+#ifdef DNNL_PPC64
+#ifdef __MMA__
```
Comment (dzarukin) [inline src/cpu/gemm/gemm.cpp:200]: QQ: why is this needed?

File: src/cpu/gemm/gemm.cpp [hunk 5]
```diff
@@ -192,6 +196,25 @@ dnnl_status_t gemm_s8x8s32(const char *transa, const char *transb,
                 B, LDB, bo, beta, C, LDC, co, false);
 #endif
 
+#ifdef DNNL_PPC64
```
Comment (dzarukin) [inline src/cpu/gemm/gemm.cpp:199]:
```suggestion
#if DNNL_PPC64
```
These macros are all defined. Only one of them is set to `1` at a time.

File: src/cpu/gemm/gemm.cpp [hunk 6]
```diff
@@ -222,6 +245,26 @@ dnnl_status_t gemm_s8x8s32(const char *transa, const char *transb,
                 LDA, ao, B, LDB, bo, beta, C, LDC, co);
 #endif
 
+#ifdef DNNL_PPC64
+#ifdef __MMA__
+    if (!(utils::one_of(*transa, 'n', 'N', 't', 'T')
+                && utils::one_of(*transb, 'n', 'N', 't', 'T')))
+        return dnnl_unimplemented;
```
Comment (dzarukin) [inline src/cpu/gemm/gemm.cpp:252]: Same concerns here.

File: src/cpu/ppc64/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+file(GLOB_RECURSE SOURCES
```
Comment (dzarukin) [inline src/cpu/ppc64/CMakeLists.txt:1]: This file needs a copyright.

File: src/cpu/ppc64/PPC64_gemm_driver.hpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
+/*******************************************************************************
+* Copyright 2018-2020 Intel Corporation
```
Comment (dzarukin) [inline src/cpu/ppc64/PPC64_gemm_driver.hpp:2]:
```suggestion
* Copyright 2022 Intel Corporation
```
Would you like to insert your copyright here instead?

File: src/cpu/ppc64/PPC64_gemm_driver.hpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef CPU_PPC64_GEMM_GEMM_DRIVER_HPP
+#define CPU_PPC64_GEMM_GEMM_DRIVER_HPP
+
+//#include "cpu/ppc64/PPC64_gemm_s8x8s32.hpp"
```
Comment (dzarukin) [inline src/cpu/ppc64/PPC64_gemm_driver.hpp:20]: Nit: Debug leftover?

File: src/cpu/ppc64/PPC64_gemm_driver.hpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace dnnl {
+namespace impl {
+namespace cpu {
+//template <typename a_type, typename b_type, typename c_type>
```
Comment (dzarukin) [inline src/cpu/ppc64/PPC64_gemm_driver.hpp:25]: Nit: Debug leftover?

File: src/cpu/ppc64/PPC64_gemm_driver.hpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+*******************************************************************************/
+
+#ifndef CPU_PPC64_GEMM_GEMM_DRIVER_HPP
+#define CPU_PPC64_GEMM_GEMM_DRIVER_HPP
+
+//#include "cpu/ppc64/PPC64_gemm_s8x8s32.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace cpu {
```
Comment (dzarukin) [inline src/cpu/ppc64/PPC64_gemm_driver.hpp:24]: One more namespace level of `ppc64` is required.

File: src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
+#ifdef __MMA__
```
Comment (dzarukin) [inline src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp:1]: Missing copyright here as well.

File: src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+#include <altivec.h>
+#include "PPC64_gemm_s8x8s32.hpp"
+#include "cpu/simple_q10n.hpp"
+//#define TIMING 1
+
+#ifdef TIMING
```
Comment (dzarukin) [inline src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp:7]: Please move all timing under verbose functionality. If it requires a special timer on PPC64, it would be fine.
Just in case, will std::chrono work?

**Review by dzarukin — COMMENTED** [review]
Thank you for this PR! Please find some comments below.

File: src/cpu/gemm/gemm.cpp (hunk 4 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp:200]: From what I understand, `DNNL_PPC64` is defined whenever we are running on any "Power" box (Power7, Power8, Power9, Power10).  However, the "MMA engine" is new with Power10, so we must also restrict access by using the environment variable `__MMA__` which is only defined for boxes which contain that feature.

**Comment by vpirogov** [comment]
@austinpagan, we don't have known problems with code dumps in gtests. This may be something specific to PPC64 or something introduced by your changes.

**Comment by austinpagan** [comment]
Thank you for confirming.  I will focus on the core dumps FIRST.

File: src/cpu/gemm/gemm.cpp (hunk 2 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp:36]: I looked in wikipedia for the definition of CamelCase, and I'm now confused.
Are you saying you prefer "ppc64_gemm_driver.hpp" over "PPC64_gemm_driver.hpp"?
If so, we can certainly do that.  I'm just trying to be sure I know what you're wanting.

File: src/cpu/gemm/gemm.cpp (hunk 5 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp]: I made that change, as it looked like it might fix some of our bugs.  It didn't, but anyway, it's now in there.

File: src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp (hunk 13 above)
Comment (austinpagan) [reply src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp:7]: I can just remove the timing stuff entirely.  That was my plan, I just forgot it...

**Comment by austinpagan** [comment]
I'm not sure what's up with these build failures?  If someone wants to tell me this is somehow a known problem, I can move on to addressing the various comments above, but until then, I am not sure how to proceed.

File: src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp (hunk 13 above)
Comment (austinpagan) [reply src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp:7]: fixed in fourth commit

File: src/cpu/ppc64/PPC64_gemm_driver.hpp (hunk 10 above)
Comment (austinpagan) [reply src/cpu/ppc64/PPC64_gemm_driver.hpp:25]: fixed in fourth commit

File: src/cpu/ppc64/PPC64_gemm_driver.hpp (hunk 9 above)
Comment (austinpagan) [reply src/cpu/ppc64/PPC64_gemm_driver.hpp:20]: fixed in fourth commit

File: src/cpu/gemm/gemm.cpp (hunk 2 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp:36]: fixed CamelCase filenames and subroutine call name in fourth commit.

**Comment by austinpagan** [comment]
So...  before I address the remaining five comments/suggestions above, could someone please help me understand about this one failing test above?  It does not look like it's failing in anything involving 8 bit matrix multiplication...

File: src/cpu/CMakeLists.txt (hunk 1 above)
Comment (austinpagan) [reply src/cpu/CMakeLists.txt:47]: I am not an expert in the use of these cmake things, but I think this is here for the single purpose of guaranteeing the compile flags "-O3 and -funroll-loops" would be added to the compilation of our code.  But thanks for pointing me at this, because we need to add more code to this list!

File: src/cpu/gemm/gemm.cpp (hunk 3 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp:203]: Moved as requested.

File: src/cpu/gemm/gemm.cpp (hunk 6 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp:252]: Same fix applied!

File: src/cpu/ppc64/CMakeLists.txt (hunk 7 above)
Comment (austinpagan) [reply src/cpu/ppc64/CMakeLists.txt:1]: Copyright added.

File: src/cpu/ppc64/PPC64_gemm_driver.hpp (hunk 8 above)
Comment (austinpagan) [reply src/cpu/ppc64/PPC64_gemm_driver.hpp]: Yes.  Done.

File: src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp (hunk 12 above)
Comment (austinpagan) [reply src/cpu/ppc64/PPC64_gemm_s8x8s32.cpp]: Copyright added.

**Comment by austinpagan** [comment]
Our fifth commit addresses the remaining comments we received above.
It passes both test suites (no core dumps, no failures)

**Comment by austinpagan** [comment]
So, that last test (continuous-integration/drone/pr) is still "Build is pending" after eight hours!

**Comment by nSircombe** [comment]
> So, that last test (continuous-integration/drone/pr) is still "Build is pending" after eight hours!

There's quite a backlog - https://cloud.drone.io/oneapi-src/oneDNN, it's working through them in serial I believe.

**Comment by austinpagan** [comment]
I just added one last Copyright statement in src/cpu/gemm/gemm.cpp, which
kicked off a new round of automated builds.

But I did some checking on that one test build that keeps failing (continuous-integration/drone/pr).
The failure is in one single test in the 99-test unit test suite.
That test (Number 71) is called test_convolution_eltwise_forward_f32.

I instrumented our code to print out a BIG MESSAGE if we ever entered either routine 
in src/cpu/gemm/gemm.cpp that calls into our new code.

I ran that test in my tree, after doing a full build with my debug printout.
The test passed, and my printout never showed up (which I suppose is not 
a surprise, since our new code has nothing to do with any float32 values).

I do NOT know why in the automated CI environment it fails.

I COULD push my printout into the main repository and see if the print out appears during your automated test,
but I'm hoping we don't have to go to such extremes?

**Comment by vpirogov** [comment]
@austinpagan, looks like this is something PPC64 specific, but not related to your PR. I believe failures in test_convolution_eltwise_forward_f32 should not block this changeset.

**Comment by austinpagan** [comment]
So, I looked at the "v2.7" page, and I can see this PR there (**_thank you for that!_**), 
but it still says "review required"...

**Comment by vpirogov** [comment]
@austinpagan, the PR still needs to be approved by @dzarukin to land into the master branch.

**Comment by austinpagan** [comment]
I appreciate the clarification.  I think I have addressed the comments/requests documented above by @dzarukin, and I stand ready to jump on it if I've missed any.

**Review by dzarukin — COMMENTED** [review]
BTW, could you, please, kindly clarify the choice of dnnl_sgemm API over matmul primitive API which has broader usability features enabled? Is there some existed integration you target with this change? Thanks.

**Comment by austinpagan** [comment]
Oh, jeez.  Clang Format.  Sorry.  Will redo.

**Comment by austinpagan** [comment]
I guess I didn't need to do a "clang-format" on the CMakeLists.txt file...

File: src/cpu/CMakeLists.txt [hunk 14]
```diff
@@ -1,20 +1,20 @@
...
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-#===============================================================================
+#Unless required by applicable law or agreed to in writing, software
+#distributed under the License is distributed on an "AS IS" BASIS,
+#WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#See the License for the specific language governing permissions and
+#limitations under the License.
+#== == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == == =
```
Comment (dzarukin) [inline src/cpu/CMakeLists.txt:17]: Could you drop the modification of copyright, please?

File: src/cpu/gemm/gemm.cpp [hunk 15]
```diff
@@ -192,6 +203,22 @@ dnnl_status_t gemm_s8x8s32(const char *transa, const char *transb,
                 B, LDB, bo, beta, C, LDC, co, false);
 #endif
 
+#if DNNL_PPC64
```
Comment (dzarukin) [inline src/cpu/gemm/gemm.cpp:206]: Kindly suggest to follow the `elif` practice as in headers.

```suggestion
#elif DNNL_PPC64
```

File: src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+        } else {
+            for (int i = 0; i < b_size; ++i)
+                Bshort[i] = ((short)B[i]) - (short)*bo;
+        }
+        APraw = (short *)malloc((m_cap * k_cap + 15) * sizeof(short), 4096);
+        BPraw = (short *)malloc((k_cap * n_cap + 15) * sizeof(short), 4096);
+        if (utils::any_null(APraw, BPraw)) {
+            free(APraw);
+            free(BPraw);
+            return dnnl_out_of_memory;
```
Comment (dzarukin) [inline src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:66]: I guess `Ashort` and `Bshort` will leak here.

Does this platform support clang compiler? If yes, did you have a chance to run oneDNN set of gtests named `test_gemm_x8x8s32`, where x is either `s` or `u` with different sanitizers to make sure that code doesn't expose any issues?

File: src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+        if (ATflag)
+            pack_N16_8bit(k, m, A, lda, AP);
+        else
+            pack_T16_8bit(k, m, A, lda, AP);
+        if (flipB_flag) {
+            int b_size = ldb * (BTflag ? k - 1 : n - 1) + (BTflag ? n : k);
+            uint8_t *Bflip = (uint8_t *)malloc(b_size * sizeof(uint8_t), 4096);
+            if (utils::any_null(Bflip)) {
+                free(Bflip);
+                return dnnl_out_of_memory;
```
Comment (dzarukin) [inline src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:106]: Same leak here.

File: src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+                pack_T8_8bit(k, n, B, ldb, BP);
+            else
+                pack_N8_8bit(k, n, B, ldb, BP);
+        }
+        gemm_kernel_8bit(m, n, k, (float)alpha, AP, BP, C, beta, ldc);
+        if (flipB_flag) {
+            int *comparray = (int *)malloc(m * sizeof(int), 4096);
+            if (utils::any_null(comparray)) {
+                free(comparray);
+                return dnnl_out_of_memory;
```
Comment (dzarukin) [inline src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:126]: And here.

File: src/cpu/CMakeLists.txt (hunk 14 above)
Comment (austinpagan) [reply src/cpu/CMakeLists.txt]: Done

File: src/cpu/gemm/gemm.cpp (hunk 15 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp]: Done

File: src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp (hunk 16 above)
Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:66]: I do not understand what you mean by "leak".  Memory leak, or some sort of overflow/underflow problem?
Also, I do not know what a "sanitizer" is.

We can compile this code using gcc or clang.

We routinely run all four of the tests you mentioned.
test_gemm_u8u8s32 and test_gemm_s8u8s32 trivially pass because those combinations are not supported.
test_gemm_u8s8s32 and test_gemm_s8s8s32 both pass all 293 tests.

Comment (dzarukin) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:66]: `Ashort` and `Bshort` are dynamically allocated. They are validated at the point of allocation and that's it. Now, if `APraw` or `BPraw` fail the allocation, then `Ashort` and `Bshort` will not be freed at this moment, since the section below only frees `APraw` and `BPraw`, but not `Ashort` and `Bshort`. This is a memory leak, basically.

"Sanitizers" are usually referred to special `-fsanitize` instrumentation [options](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html). oneDNN supports these options internally through build flags: `-DDNNL_USE_CLANG_SANITIZER=<sanitizer_name>` where `<sanitizer_name>` follows the option value name like "Address" or "Undefined". Once built with these options with clang compiler, runtime will do special instrumentation and might find additional issues which hard to find with manual code review.

File: src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp (hunk 18 above)
Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:132]: My apologies.  Thank you for your patience in explaining this to me.
I have dropped code that I hope you will agree fixes this memory leak issue.

File: src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp (hunk 16 above)
Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: So, I looked at the page that you pointed us at, and I see ten options for -fsanitize:
   address
   kernel-address
   hwaddress
   kernel-hwaddress
   pointer-compare
   pointer-subtract
   shadow-call-stack
   thread
   leak
   undefined
   
Also, there are 22 sub-options under the "undefined" option.

Question 1: which ones would you like us to test?

Question 2: is it OK if we just use the `-fsanitize=[option]` construction rather than the `-DDNNL_USE_CLANG_SANITIZER` construction, so that we can only apply the build option to our code rather than all of OneDNN?

Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: @dzarukin see above...

Comment (dzarukin) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: My recommendation to focus on Address one and Undefined one. Those will cover most errors if anything pops up. They also include some sub-options in them.

You should be fine to enable it for the whole project, there shouldn't be any (at least, the we were aware of) issues associated with the piece you are modifying. Thank you.

Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: Status update:
First, THANK YOU for your guidance!  Much appreciatedl
As it happens, there was a bug in the Clang compiler for PPC64, but our team has produced a fix which will be eventually promoted into the official distribution.
With this fix, we are able to compile and run all of OneDNN using Clang on our Power10 processor, and it passes all tests.
More importantly, we do see that when we add the sanitizer flag (specifying Address), we find three failures in the "make test" suite, which are clearly in my code.  I'm running the various "make test_benchdnn_*" options now to see how many errors there are there.  My work continues, and I will keep you informed!

Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: In all the test_benchdnn_* tests, I only uncovered one failure, when I specified "Address." I found and fixed two bugs and now both test suites run clean. I am going to compile with "Undefined" now.

Comment (austinpagan) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: Update: found ONE more bug, hope to have it tested and passing all test suited within 24 to 48 hours.
When is the cutoff?

Comment (dzarukin) [reply src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp:68]: Feel free to take your time to resolve issues. We appreciate higher quality, and will backport to the version needed. Thank you.

File: src/cpu/gemm/gemm.cpp [hunk 19]
```diff
@@ -222,6 +247,23 @@ dnnl_status_t gemm_s8x8s32(const char *transa, const char *transb,
...
+    int ATflag = (*transa == 'T') || (*transa == 't');
+    int BTflag = (*transb == 'T') || (*transb == 't');
+    int m = (int)*M;
+    int n = (int)*N;
+    int k = (int)*K;
+    int lda = (int)*LDA;
+    int ldb = (int)*LDB;
+    int ldc = (int)*LDC;
+
+    return cblas_gemm_s8u8s32_ppc64(ATflag, BTflag, offsetc, m, n, k, *alpha, A,
```
Comment (mgouicem) [inline src/cpu/gemm/gemm.cpp:261]: Are you sure this would work as expected?
In this function, B is of type `const int8_t *` and you cast it to `const uint8_t*`.
I would expect that you will get wrong results for some values. See this example with [godbold](https://godbolt.org/z/x3vocfoxh).

File: src/cpu/ppc64/ppc64_gemm_driver.hpp [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+#define CPU_PPC64_GEMM_GEMM_DRIVER_HPP
+
+namespace dnnl {
+namespace impl {
+namespace cpu {
+namespace ppc64 {
+dnnl_status_t cblas_gemm_s8u8s32_ppc64(int, int, char const *, int, int, int,
+        float, signed char const *, int, signed char const *,
+        unsigned char const *, int, unsigned char const *, int *, float, int,
+        int const *, int);
```
Comment (mgouicem) [inline src/cpu/ppc64/ppc64_gemm_driver.hpp:27]: I would highly recommend to use fixed size datatypes (`int8_t`, `uint8_t`, ...)

File: src/cpu/gemm/gemm.cpp (hunk 19 above)
Comment (austinpagan) [reply src/cpu/gemm/gemm.cpp:261]: My apologies.
I should have been more generous with my comments.

I chose to have one routine that handles both signed and unsigned B.
I manage that through the use of a flag at the end of the call list.

The called routine knows that if the flag is 1, the B matrix actually contains signed values, 
not unsigned, and it processes it accordingly.

At the very least, I will add comments to clarify this.

OR, I *could* replace this routine with two separate routines to handle signed and unsigned versions of B explicitly, but the flag is fairly easy to understand when you know to look for it.

Inside src/cu/ppc64/ppc64_gemm_s8x8s32.cpp, look for "flipB_flag".

If you're wanting me to create two routines rather than one with a flag, let me know and I'll do that, but my default plan now is to add comments here to explain the behavior.

**Comment by austinpagan** [comment]
I owe you one more commit with the added comments and the fixed size datatypes (as described above), after I receive confirmation that these changes will suffice.

**Comment by austinpagan** [comment]
@mgouicem : I implemented the comment, changed the name of the routine to "...s8x8s32...", implemented the int8_t, the uint8_t, the dim_t, and did some things in the BIG file to get it to compile with these new changes.  It all compiles and runs, and again (assuming it passes the automated tests), I would say my code is ready for you.

File: src/cpu/CMakeLists.txt [hunk 21]
```diff
@@ -38,6 +38,14 @@ if((DNNL_TARGET_ARCH STREQUAL "X64") OR (DNNL_TARGET_ARCH STREQUAL "AARCH64"))
     endforeach()
 endif()
 
+if(DNNL_TARGET_ARCH STREQUAL "PPC64")
+    file(GLOB FILES_REQUIRED_OPT
+        ${CMAKE_CURRENT_SOURCE_DIR}/gemm/*.[ch]pp
+    )
+    set_source_files_properties(${FILES_REQUIRED_OPT}
+        PROPERTIES COMPILE_FLAGS "-O3 -funroll-loops")
```
Comment (mgouicem) [inline src/cpu/CMakeLists.txt:46]:
```suggestion
    if(NOT UPPERCASE_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
        set_source_files_properties(${FILES_REQUIRED_OPT}
            PROPERTIES COMPILE_FLAGS "-O3 -funroll-loops")
    endif()
```

Maybe guard this so that optimizations are not applied in debug build?

**Review by mgouicem — APPROVED** [review]
Thanks for your contribution!
This looks good to me!

**Comment by austinpagan** [comment]
@dzarukin you had mentioned that it was curious we were not focusing on the "matmul" primitive directly, and I had explained our thought process on that, saying that we could investigate doing some PPC64-specific optimizations on matmul as well. 

Assuming this pull request gets approved, that will be the next thing I do, and I am starting to investigate that interface now...

**Comment by austinpagan** [comment]
I have good news, regarding the matmul primitive API.

Early on, I had focused on the "gemm_s8x8s32" call within the src/cpu/gemm directory.
From what I could tell, that was the choke point.  all in8 matrix multiplication went through this call.

I have confirmed two things recently:

If you're doing float32 matrix multiplication, either through dnnl_sgemm or matmul, you are calling cblas_sgemm, so you are already getting our MMA-accelerated float32 code (in our OpenBLAS library) if you run on Power.

If you're doing int8 matrix multiplication, either through dnnl_gemm or matmul, you are calling gemm_s8x8s32 within oneDNN, and THAT is the code we have optimized in this pull request.

So, for float32 and int8, all you have to do is accept this pull request, and the matmul primitive API will access our MMA-enhanced code on Power.

**Comment by austinpagan** [comment]
I have an update to this pull request that I will be committing in the next few days.

**Comment by austinpagan** [comment]
@dzarukin 
We have code that accelerates bfloat16 matrix multiplication when running on Power10.
I have now coded up a couple changes in my working branch of oneDNN that give users of matmul access to this accelerated code, and will drop that code as a subsequent pull request as soon as this one is approved and promoted into the repository.

**Comment by austinpagan** [comment]
@dzarukin Have I missed something you were expecting me to do?

**Comment by dzarukin** [comment]
Thank you for the contribution!

**Comment by austinpagan** [comment]
Glad you like it.
We will create a new pull request soon (as I indicated above) to enable access to our OpenBLAS bfloat16 code, for when OneDNN wants to do matrix multiplication of 16-bit floats in that particular format.
